include/gatekeeper_mailbox.h

```diff
@@ -19,14 +19,19 @@
 #ifndef _GATEKEEPER_MAILBOX_H_
 #define _GATEKEEPER_MAILBOX_H_
 
-/*
- * TODO Remove this preprocessing directive when there are
- * are calls to these functions; for now, the mailbox is
- * just an example of how to set up a library.
- */
-#if 0
-void create_mailbox(void);
-void destroy_mailbox(void);
-#endif
+#include <rte_ring.h>
+#include <rte_mempool.h>
+
+struct mailbox {
+	struct rte_ring    *ring;
+	struct rte_mempool *pool;
+};
+
+struct mailbox *create_mailbox(
```

Comment:
Yes to the first question.

Right now, `struct gk_instance` has the field `struct mailbox *mb`. So to reach a field of `mb`, one has to go to an instance of `struct gk_instance`, read the address held in `mb`, and look up the memory in that address. Notice that the reader has already loaded the instance of `struct gk_instance` into cache. Thus, if the field `mb` of `struct gk_instance` were `struct mailbox mb`, the second lookup wouldn't be necessary.